frontend/src/metabase-lib/extractions.ts

```diff
@@ -9,10 +36,39 @@ export function columnExtractions(
   return ML.column_extractions(query, column);
 }
 
-export function extract(
+export type ColumnExtractionTag =
+  | "hour-of-day"
+  | "day-of-month"
+  | "day-of-week"
+  | "month-of-year"
+  | "quarter-of-year"
+  | "year"
+  | "domain"
+  | "host"
+  | "subdomain";
+
+/**
+ * Return the functions used by a specific column extraction.
+ */
+export function functionsUsedByExtraction(
   query: Query,
   stageIndex: number,
   extraction: ColumnExtraction,
-): Query {
-  return ML.extract(query, stageIndex, extraction);
+): string[] {
+  const expression = extractionExpression(query, stageIndex, extraction);
+  const parts = expressionParts(query, stageIndex, expression);
+  return walk(parts);
+}
+
+function walk(parts: ExpressionParts): string[] {
+  const res: string[] = [parts.operator];
+  parts.args.forEach(arg => {
+    if (!arg || !(typeof arg === "object")) {
+      return;
+    }
+    if ("operator" in arg) {
```

Comment:
`if (`typeof arg === "object") walk(...)` without the first if

Reply:
https://github.com/metabase/metabase/blob/4b71e9aa645117a944af409882bda33f64225bf9/frontend/src/metabase-lib/filter.ts#L570

Reply:
because `typeof null === "object"`

Reply:
feel free to move this utility to a different place